### ISSUE TITLE
Feature request: extending dynamic state

### ISSUE THREAD

**Issue by adrian17** [gpuweb/gpuweb#4014]
Tldr: similar to Vulkan's recent revisions, it'd be nice to be able to reduce overhead of creating pipelines, "pipeline combinatorial explosion" and recreating them in some applications, by allowing more 
hopefully-commonly-supported dynamic state, like depth test or blend mode.

The original research of pipeline state design in https://github.com/gpuweb/gpuweb/issues/26 is now slightly out of date: the set of originally supported dynamic state fields had 9 items, but by Vulkan 1.3 it has several dozen, including common items like stencil or depth compare test operation. This is provided by now-required extensions [VK_EXT_extended_dynamic_state](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_extended_dynamic_state.html) and [VK_EXT_extended_dynamic_state2](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_extended_dynamic_state2.html). The next (currently optional) extension [VK_EXT_extended_dynamic_state3](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_extended_dynamic_state3.html) adds more dynamic state like blend mode. This admittedly complicates the API by having two ways of doing the same thing (and having the implementation track these two ways), but clearly they considered it worth it given they are being slowly moved into the core spec.

( digression: there are also other extensions like (required in 1.3) [VK_KHR_dynamic_rendering](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_dynamic_rendering.html) that reduced focus on RenderPass objects, and a very recent (optional) [VK_EXT_shader_object](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_shader_object.html) that essentially abandoned pipelines and re-added dynamic shader binding. In particular, the last one's proposal contains a [rationale](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_EXT_shader_object.adoc#problem-statement) claiming the pipeline design to have - in hindsight - failed. )

Our application (which works as a game engine and can receive state changes at runtime - this can't be redesigned) is also hitting the same kind of issues relating to the number of pipelines bloating application load time, and some applications requiring us to recompile pipelines at runtime. Being able to change some of this state - like blend mode or depth test - dynamically would let us mitigate that significantly. And from what I understand, at least some of this dynamic state should be widely supported across vendors.

What's your opinion on this kind of API extension?

**Comment by kainino0x** [comment]
Anything that starts to see significant uptake across the industry, and can be done on at least 2 of Metal/D3D12/Vulkan, is fair game! I don't know if we've started seeing this in Metal or D3D12 yet though, I haven't heard anything about it.
https://github.com/gpuweb/gpuweb/blob/main/process/RequirementsForAdditionalFunctionality.md

**Comment by adrian17** [comment]
> I don't know if we've started seeing this in Metal or D3D12 yet though, I haven't heard anything about it.

Metal already had some of these dynamic, see https://github.com/gpuweb/gpuweb/issues/26

> States that are surprisingly out of PSO:
> - depth bias
> - cull mode
> - triangle fill mode
> - depth/stencil

And a couple more, like front face winding order. See https://developer.apple.com/documentation/metal/mtlrendercommandencoder .

**Comment by mr-mobster** [comment]
Just a quick note on this:

> But indeed, D3D12 seem to keep them in PSO. (weirdly, it seems to support dynamic primitive topology, while in Metal it's pipeline-bound.)

It's only pipeline-bound in Metal if you use [layered rendering](https://developer.apple.com/documentation/metal/render_passes/rendering_to_multiple_texture_slices_in_a_draw_command#3362353):

>  Render pipelines that can render to layers must specify the type of primitive that they can render. This requirement differs from a normal render pipeline, where you can select a different primitive for each draw command.
